### PR TITLE
OS-2018 - create workload instance with an empty payload secret uuid

### PR DESCRIPTION
**Pull request by SuhasTA4**
image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,7 +26,7 @@ Repo for automation build, test etc.
     - CentOS 8
          - Install required packages
            ```
-           yum install gcc python3-virtualenv epel-release centos-release-openstack-train -y
+           yum install gcc python3-virtualenv epel-release centos-release-openstack-wallaby python3-barbicanclient -y
```
Comment (savitha-p) [inline README.md:29]: you need to add barbicanclient dependency for ubuntu based tempest vms as well

File: tempest/api/workloadmgr/barbican/test_barbican.py [hunk 2]
```diff
@@ -3346,3 +3346,83 @@ def test_13_cleanup(self):
...
+            #create key pair...
+            self.kp = self.create_key_pair(tvaultconf.key_pair_name)
+
+            #create vm...
+            self.vm_id = self.create_vm(key_pair=self.kp)
+
+            #create secret with empty payload.
+            empty_payload = "\"\""
+            secret_key_cmd = command_argument_string.openstack_create_secret + \
+                                  "-p " + str(empty_payload) + \
```
Comment (savitha-p) [inline tempest/api/workloadmgr/barbican/test_barbican.py]: -p needs to be part of command_argument_string only.

File: tempest/api/workloadmgr/barbican/test_barbican.py [hunk 3]
```diff
@@ -3346,3 +3346,83 @@ def test_13_cleanup(self):
...
+                                  " --insecure"
```
Comment (savitha-p) [inline tempest/api/workloadmgr/barbican/test_barbican.py]: Insecure tag needs to be handled similar to wlm cli commands.

File: tempest/api/workloadmgr/barbican/test_barbican.py [hunk 4]
```diff
@@ -3346,3 +3346,83 @@ def test_13_cleanup(self):
...
+                    tvaultconf.FAIL)
+                reporting.set_test_script_status(tvaultconf.FAIL)
+                raise Exception("Create secret with an empty payload.")
+            else:
+                reporting.add_test_step(
+                    "Create secret with an empty payload.",
+                    tvaultconf.PASS)
+
+            #parse the output to get the secret uuid.
+            out = cli_parser.cli_output(secret_key_cmd)
```
Comment (savitha-p) [inline tempest/api/workloadmgr/barbican/test_barbican.py:3375]: In positive scenario, 2 secrets would get created in this test , but are not getting cleaned up. Need to handle cleanup as well.

File: tempest/api/workloadmgr/barbican/test_barbican.py [hunk 5]
```diff
@@ -3346,3 +3346,83 @@ def test_13_cleanup(self):
...
+                                  "-p " + str(empty_payload) + \
+                                  " --insecure"
+
+            error = cli_parser.cli_error(secret_key_cmd)
+            if error and (str(error.strip('\n')).find('ERROR') != -1):
+                reporting.add_test_step(
+                    "Create secret with an empty payload.",
+                    tvaultconf.FAIL)
+                reporting.set_test_script_status(tvaultconf.FAIL)
+                raise Exception("Create secret with an empty payload.")
```
Comment (savitha-p) [inline tempest/api/workloadmgr/barbican/test_barbican.py]: When Exception is raised, adding test step with failed status and setting test script status to FAIL is already handled as part of Exception block. These repetitive code can be removed

File: tempest/api/workloadmgr/barbican/test_barbican.py [hunk 6]
```diff
@@ -3346,3 +3346,83 @@ def test_13_cleanup(self):
...
+            else:
+                reporting.add_test_step(
+                    "Create secret with an empty payload.",
+                    tvaultconf.PASS)
+
+            #parse the output to get the secret uuid.
+            out = cli_parser.cli_output(secret_key_cmd)
+            LOG.debug("Response from CLI: " + str(out))
+
+            out = out.split('/secrets/')[1].split('|')[0].strip('.')
```
Comment (savitha-p) [inline tempest/api/workloadmgr/barbican/test_barbican.py]: You can add the command as `openstack secret store -p "" -f json` and dump the response to json object , which will be easy to fetch further details from response.

File: tempest/api/workloadmgr/barbican/test_barbican.py [hunk 7]
```diff
@@ -3346,3 +3346,83 @@ def test_13_cleanup(self):
...
+                    if (result):
+                        reporting.add_test_step("Creation of workload instance cannot have an empty payload",
+                                tvaultconf.PASS)
+                        reporting.set_test_script_status(tvaultconf.PASS)
+                    else:
+                        reporting.add_test_step("Different error occurred.",
+                                tvaultconf.FAIL)
+                        reporting.set_test_script_status(tvaultconf.FAIL)
+            else:
+                LOG.debug("workload with encryption created successfully")
```
Comment (savitha-p) [inline tempest/api/workloadmgr/barbican/test_barbican.py:3411]: In this case, workload creation should be handled.

File: tempest/command_argument_string.py [hunk 8]
```diff
@@ -45,6 +45,9 @@
 restore_security_groups = "workloadmgr restore-security-groups "
 network_restore = "workloadmgr restore-network-topology "
 
+#openstack secret commands
+openstack_list_secret = "openstack secret list"
```
Comment (savitha-p) [inline tempest/command_argument_string.py]: Remove any unused variables

File: tempest/api/workloadmgr/barbican/test_barbican.py (hunk 4 above)
Comment (SuhasTA4) [reply tempest/api/workloadmgr/barbican/test_barbican.py:3375]: Now we are focusing directly on cli_output and cleaning-up the secret key at the end of test case.

File: tempest/api/workloadmgr/barbican/test_barbican.py (hunk 2 above)
Comment (SuhasTA4) [reply tempest/api/workloadmgr/barbican/test_barbican.py]: Handled in command_argument_string.py

File: tempest/command_argument_string.py [hunk 9]
```diff
@@ -45,6 +45,9 @@
 restore_security_groups = "workloadmgr restore-security-groups "
 network_restore = "workloadmgr restore-network-topology "
 
+#openstack secret commands
+openstack_create_secret_with_empty_payload = "openstack secret store -p \"\" -f json --insecure "
```
Comment (savitha-p) [inline tempest/command_argument_string.py]: Insecure flag should not be added to the command here. Instead it should be appended during execution based on AUTH_URL. This should be handled in fetch_resources.sh

**Review by savitha-p — APPROVED** [review]

**Review by nasharatm — APPROVED** [review]

**Pull request merged**